File: LeetCode/1358. Number of Substrings Containing All Three Characters/solution.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        cnt = collections.defaultdict(int)
        res, left, n = 0, 0, len(s)
        for right in range(n):
            cnt[s[right]] += 1
            while left < n and sum(list(map(lambda x: x>0, cnt.values()))) == 3:
                res += n-right
                cnt[s[left]] -= 1
                left += 1
        return res
                
        
class Solution:
    # def numberOfSubstrings(self, s):
    #     res = i = 0
    #     count = {c: 0 for c in 'abc'}
    #     for j in range(len(s)):
    #         count[s[j]] += 1
    #         while all(count.values()):
    #             count[s[i]] -= 1
    #             i += 1
    #         res += i
    #     return res

    def numberOfSubstrings(self, s):
        res, last = 0, [-1] * 3
        for i, c in enumerate(s):
            last[ord(c) - 97] = i
            res += 1 + min(last)
        return res

File: LeetCode/1358. Number of Substrings Containing All Three Characters/solution.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        vector<int> cc(3,0);
        int cnt = 0, left = 0, n = s.length(), res = 0;
        for (int right = 0; right < n; right++){
            int idx = s[right]-'a';
            if (cc[idx] == 0)
                cnt++;
            cc[idx]++;
            while (cnt == 3){
                int idx = s[left]-'a';
                cc[idx]--;
                if(cc[idx] == 0)
                    cnt--;
                left++;
            }
            res += left;
        }
        return res;

    }
};